src/win/scandir.c

```diff
@@ -0,0 +1,145 @@
+#include <stdlib.h>
+#include <stddef.h>
+#include <string.h>
+#include "scandir.h"
+
+int alphasort(const struct dirent** dirent1, const struct dirent** dirent2)
+{
+  return (strcmp((*(const struct dirent **)dirent1)->d_name,
+                 (*(const struct dirent **)dirent2)->d_name));
+}
+
+/* Initial guess at directory allocated.  */
+#define INITIAL_ALLOCATION 20
+
+int scandir(const char *directory_name,
+            struct dirent ***array_pointer,
+            int (*select_function) (const struct dirent *),
+            int (*compare_function) (const struct dirent**, const struct dirent**))
+{
+  DIR *directory;
+  struct dirent **array;
+  struct dirent *entry;
+  struct dirent *copy;
+  int allocated = INITIAL_ALLOCATION;
+  int counter = 0;
+
+  /* Get initial list space and open directory.  */
+  directory = opendir(directory_name);
+  if(directory == NULL) return -1;
+
+  array = (struct dirent **) malloc(allocated * sizeof(struct dirent *));
+  if(array == NULL) return -1;
+
+  /* Read entries in the directory.  */
+
+  while(entry = readdir(directory), entry)
+    if(select_function == NULL || (*select_function) (entry))
+    {
+      /* User wants them all, or he wants this one.  Copy the entry.  */
+
+      /*
+       * On some OSes the declaration of "entry->d_name" is a minimal-length
+       * placeholder.  Example: Solaris:
+       *         /usr/include/sys/dirent.h:
+       *                "char d_name[1];"
+       *        man page "dirent(3)":
+       *                The field d_name is the beginning of the character array
+       *                giving the name of the directory entry. This name is
+       *                null terminated and may have at most MAXNAMLEN chars.
+       * So our malloc length may need to be increased accordingly.
+       *        sizeof(entry->d_name): space (possibly minimal) in struct.
+       *         strlen(entry->d_name): actual length of the entry. 
+       *
+       *                        John Kavadias <john_kavadias@hotmail.com>
+       *                        David Lee <t.d.lee@durham.ac.uk>
+       */
+      int namelength = strlen(entry->d_name) + 1;        /* length with NULL */
+      int extra = 0;
+
+      if(sizeof(entry->d_name) <= namelength)
+      {
+        /* allocated space <= required space */
+        extra += namelength - sizeof(entry->d_name);
+      }
+
+      copy = (struct dirent *) malloc(sizeof(struct dirent) + extra);
+      if(copy == NULL)
+      {
+        closedir(directory);
+        free(array);
+        return -1;
+      }
+      copy->d_ino = entry->d_ino;
+      copy->d_reclen = entry->d_reclen;
+      strcpy(copy->d_name, entry->d_name);
+
+      /* Save the copy.  */
+
+      if(counter + 1 == allocated)
+      {
+        allocated <<= 1;
+        array = (struct dirent **)
+          realloc((char *) array, allocated * sizeof(struct dirent *));
+        if(array == NULL)
+        {
+          closedir(directory);
+          free(array);
```

Comment:
in this path, array is null => not needed and will crash. Only `g_free()` cope with NULL pointer.

Reply:
free shouldn't crash on NULL, but anyway it doesn't make sense to call it here at all

Reply:
Indeed, wrong recollection I thought it was a diff between g_free/free but it is not. My confusion comes from the fact that most `g_...()` routines working on strings can take a NULL pointer whereas the equivalent std function do not.